src/diffusers/schedulers/scheduling_utils.py

```diff
@@ -24,20 +23,21 @@
 
 SCHEDULER_CONFIG_NAME = "scheduler_config.json"
 
-
-class KarrasDiffusionSchedulers(Enum):
-    DDIMScheduler = 1
-    DDPMScheduler = 2
-    PNDMScheduler = 3
-    LMSDiscreteScheduler = 4
-    EulerDiscreteScheduler = 5
-    HeunDiscreteScheduler = 6
-    EulerAncestralDiscreteScheduler = 7
-    DPMSolverMultistepScheduler = 8
-    DPMSolverSinglestepScheduler = 9
-    KDPM2DiscreteScheduler = 10
-    KDPM2AncestralDiscreteScheduler = 11
-    DEISMultistepScheduler = 12
+# NOTE keep in sync with ./__init__.py `KarrasDiffusionSchedulers`
+karras_diffusion_scheduler_compatibles = [
+    "DDIMScheduler",
+    "DDPMScheduler",
+    "PNDMScheduler",
+    "LMSDiscreteScheduler",
+    "EulerDiscreteScheduler",
+    "HeunDiscreteScheduler",
+    "EulerAncestralDiscreteScheduler",
+    "DPMSolverMultistepScheduler",
+    "DPMSolverSinglestepScheduler",
+    "KDPM2DiscreteScheduler",
+    "KDPM2AncestralDiscreteScheduler",
+    "DEISMultistepScheduler",
+]
```

Comment:
I understand the `Union` semantics, but not a big fan of duplicating these in both `KarrasDiffusionSchedulers` and here (despite the warning comments). We can maybe work around the cyclic references by using forward type declarations in the `Union`, but I don't think that would help here anyway. Not sure how to deal with this, honestly.

Reply:
I don't think we can work around the cyclic references with forward declarations because the forward declaration must still evaluate to correct code once the module with the forward declaration is fully loaded. I think they only really work when the type being forward declared is already in the same module as the usage of the type.

```
The string literal should contain a valid Python expression (i.e., compile(lit, '', 'eval') should be a valid code object) and it should evaluate without errors once the module has been fully loaded. The local and global namespace in which it is evaluated should be the same namespaces in which default arguments to the same function would be evaluated.
```
https://peps.python.org/pep-0484/#forward-references:~:text=The%20string%20literal,would%20be%20evaluated.


tl;dr we'll basically have to have regardless

```
# in <utils module>
from <scheduler module> import <a scheduler>, <b scheduler>

# can be forward reference or not, but if is forward reference, must have <a scheduler> and <b scheduler> in its module by the time its module is fully loaded
TheSchedulersUnion = Union[<a scheduler>, <b scheduler>] 
```

```
# in scheduler module
from <utils module> import TheSchedulersUnion # <- circular reference

<do something with TheSchedulersUnion>
```

The only way the Unions + forward references could work (afaik) would be

```python

from typing import Union

TheSchedulersUnion = Union['SchedulerA', 'SchedulerB']

class SchedulerA:
    prop: TheSchedulersUnion

class SchedulerB:
    prop: TheSchedulersUnion
```

But once they're in different modules, there's going to be circular imports

Reply:
I think duplicating might be the only way to use the correct union type. Also, I don't know if there's a good way to extract the values out of union. `Union.__args__` returns the types it's constructed with but I had to find that by reading the source code of Union and I wouldn't want to use undocumented features of typing.py